Enable Windows build using Visual Studio 2015

peter-b opened the pull request:
This PR builds on #5134 and adds some additional patches to the build infrastructure to enable building on an actual Windows host.

N.b. this must be merged with livecode/livecode-thirdparty#91.

peter-b committed:
lc-compile: Always define a yywrap() function

Since we never link in libfl on any platform, unconditionally
define a `yywrap()` function.  This allows compilation with flex
2.6.

runrevmark committed:
[[ VS2015 ]] Restructure Windows prebuilts

This patch reworks both the build files and names of windows
prebuilt libraries for compatibility with VS2015.

The main change is that the type and version of CRT used to build
the library is now part of the name of the prebuilt. This is due
to the fact that VS2015 will not allow you to link together static
libraries which use a mixture of Debug and Release CRTs.

Each prebuilt library now installs into its own folder under
the prebuilt/unpacked folder. The relevant gyp files in this
repo and in thirdparty have been updated to reflect that for
windows only.

The fetch-libraries script has had windows-specific actions added
including a hard-coded exception for CEF which (as it contains no
static libraries) has no dependence on CRT or visual studio version.

The build-libs batch script can be called with a single parameter
to indicate one of openssl, curl or icu to be built. The build uses
the ARCH, MODE and TOOL environment variables to determine the build.
For example ARCH=x86, MODE=debug and TOOL=14 will build a prebuilt
using the VS2015 toolchain (which is v140), the static debug CRT (/MTd
option) for x86.

There is a new batch script build-all-libs which builds four variants
of a library simultaneously - debug/x86, debug/x86_64, release/x86 and
release/x86_64. It takes a single parameter indicating which library
to build.

Note: ICU does not like the parallelization used for build-all-libs so
ICU must be built individually using build-libs and setting env vars.

runrevmark committed:
[[ VS2015 ]] Remove dependence on QuickTime

This patch removes the build systems dependence on the QuickTime
SDK being installed, and removes the use of QuickTime on Windows
from the revvideograbber component.

runrevmark committed:
[[ VS2015 ]] Remove dependence on SAPI4

This patch removes revspeech's dependence on SAPI4.

runrevmark committed:
[[ VS2015 ]] Add export definition mode to list_stub_symbols

This patch adds an extra parameter --exportdef to the
list_stub_symbols utility.

When this option is present, the utility will generate a Windows
export definition file suitable for use to build a shared library.

It is used to build revsecurity, by generating a def file which
contains only the exports required by the weak linking for
our uses of OpenSSL.

runrevmark committed:
[[ VS2015 ]] Correct definition of constexpr for VS2015

This patch ensures that constexpr is not defined as nothing when
building using VS2015 or higher as compilers in those versions
support the keyword.

runrevmark committed:
[[ VS2015 ]] Silence MASM warnings which cause build failure

Ensure that any warnings produced by MASM (used to build libffi)
do not cause a build failure.

This patch includes a tweak to gyp to so that the WarningLevel
option can be specified for the Masm tool.

runrevmark committed:
[[ VS2015 ]] Upgrade ICU to version 58.2

This patch changes the version of ICU to use to 58.2 which supports
building with VS2015 out of the box.

runrevmark committed:
[[ VS2015 ]] Tweak foundation-auto.h for building with VS2015

The VS2015 compiler is more picky about friend and inline keywords
being used. This patch removes unnecessary use of them from the
MCAutoValueRefBase class.

runrevmark committed:
[[ VS2015 ]] Tweak use of debug CRT and Rtl functions

The latest version of the VS debug CRT includes definitions for
a previously internal function _VCrtDbgReportA so the explicit
extern has been removed.

In reverse, the RtlCaptureStackBackTrace function is no longer
defined in the environment so an explicit extern has been added.

runrevmark committed:
[[ VS2015 ]] Tweak environment options for VS2015

This patch tweaks various environment flags and options to ensure
a build works under VS2015.

It ensures that the ARCH variable is always set and defaults to x86.
It maps ARCH to the Microsoft architecture name (X86 = x86 and X64 =
x86_64) and ensures that it is used to set the target_arch of the build.

To ensure that things build consistently targetting the correct
architecture, it ensures that /MACHINE is set for the link and
lib tools.

Finally it ensures the gyp is called targetting VS 2015 output and
the architecture specified by ARCH.

runrevmark committed:
[[ VS2015 ]] Ensure configure.bat doesn't terminate the caller

This patch ensures that when configure.bat exits it does so in a
way which means the host shell is not exited.

runrevmark committed:
[[ VS2015 ]] Silence abundant warnings when building

This patch disables three specific warnings generated whilst
building:

- 4577: exception handling mode mismatch
- 4800: performance warning about cast to bool
- 4244: possible loss of data due to int truncation

These need to be fixed properly at some point, however disabling
them for now vastly reduces the build log size (making it easier
to read) and decreases build time.

peter-b committed:
Merge branch 'fix-vs2015' into peter-b/test-vs2015

Conflict in configure.bat; reworked to call `config.py` script.

peter-b committed:
build: Update config.py to support new Windows builds

- Default to MSVS 2015
- For `win-x86_64` platform, need to default to `x64` as target
  architecture
- Pass `-Dtarget_arch` to gyp for Windows configuration

peter-b committed:
build: Detect early if prebuild fetching fails

When downloading prebuilts using `curl`, pass the `--fail` flag so
that `curl` fails on an HTTP error rather than creating a broken
file containing the HTML error page sent by the remote server.

peter-b committed:
build: Fake "Fast" configuration prebuilts for win

The Windows prebuilts are named by build configuration, rather than
which runtime they were compiled with.  Both the "Fast"
configuration (used only by CI builds) and the "Release"
configuration use the release CRT rather than the debug CRT.  Fake
prebuilts for the "Fast" configuration by just copying the
"Release" prebuilts.

peter-b committed:
build: Add new make.cmd batch script for VS2015

This patch replaces the existing `make.cmd` batch script used for
building LiveCode on Windows with a new one that's designed to work
properly with the tools and SDKs installed by the VS2015 build
tools installer.

peter-b committed:
docs: Update build instructions for Windows

livecodeali commented on config.py:
```diff
@@ -504,10 +509,12 @@ def configure_android(opts):
     args = core_gyp_args(opts) + ['-Gmsvs_version=' + opts['WIN_MSVS_VERSION']]
-    args += gyp_define_args(opts, {'ms_speech_sdk4': 'MS_SPEECH_SDK4',
+    args += gyp_define_args(opts, {'target_arch':    'TARGET_ARCH',
+                                   'ms_speech_sdk4': 'MS_SPEECH_SDK4',
```
I think this line can be removed, as the dependency on SAPI4 is removed earlier in this PR

peter-b committed:
build: Remove SAPI4 support from config.py script

LiveCode no longer builds revspeech using SAPI4.

peter-b commented:
@livecodeali Thanks, I've addressed that.

livecodeali commented:
@livecode-vulcan review ok e294f29

livecode-vulcan commented:
💙 review by @livecodeali ok e294f29b69bf7e3eb239c874ab114c7bb8740fe0

livecode-vulcan commented:
😞 test failure e294f29b69bf7e3eb239c874ab114c7bb8740fe0

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/973)
* try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/979)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/971)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/972)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/979)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/993)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/970)

livecodeali committed:
[[ ICU 58.2 ]] Remove no longer existing icu prebuilt linkage

livecodeali commented:
Depends on https://github.com/livecode/livecode/pull/5206

peter-b committed:
Make CEF subprocess helpers explicitly depend on ICU

livecodeali commented:
@livecode-vulcan retry e294f29

peter-b committed:
build: Make sure "x86" finds "i386" prebuilts

x86 Linux build depend on the existence of "x86" prebuilt libraries,
but the internal naming used by the prebuilt system is "i386".  It
seems that commit 48e72cf53acd may have broken things; this makes it
explicit that "i386" to "x86" renaming should be done on all
platforms.

peter-b committed:
build: Don't attempt to collect minimized ICU data

We have not yet compiled minimized ICU data for ICU 58.2.  Rejig
the ICU build script to just use the full data.  See also
http://quality.livecode.com/show_bug.cgi?19198

peter-b committed:
build: Fix typo in fetch-libraries script

In commit e7abec85fb3a, `wget` was typoed to `get`, breaking the
prebuilt library fetching for the Emscripten engine build.

peter-b committed:
Merge branch 'develop' into fix-vs2015

livecodeali commented:
@livecode-vulcan review ok fb5499e

livecode-vulcan commented:
💙 review by @livecodeali ok fb5499e64cd921103c61bbfec94169a29dad4b22

livecode-vulcan commented:
😎 test success fb5499e64cd921103c61bbfec94169a29dad4b22

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1052)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1055)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1047)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1053)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1057)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1069)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1047)
